test/integration_test.go

```diff
@@ -747,6 +751,109 @@ func TestTerragruntOutputAllCommandSpecificVariableIgnoreDependencyErrors(t *tes
+
+	// forces plugin download & initialization (no parallelism control)
+	runTerragrunt(t, fmt.Sprintf("terragrunt plan-all --terragrunt-non-interactive --terragrunt-working-dir %s -var url=%s", environmentPath, serverUrl))
+	runTerragrunt(t, fmt.Sprintf("terragrunt apply-all --terragrunt-parallelism %d --terragrunt-non-interactive --terragrunt-working-dir %s -var url=%s", parallelism, environmentPath, serverUrl))
+
+	var (
+		stdout bytes.Buffer
+		stderr bytes.Buffer
+	)
+	// Call runTerragruntCommand directly because this command contains failures (which causes runTerragruntRedirectOutput to abort) but we don't care.
+	err := runTerragruntCommand(t, fmt.Sprintf("terragrunt output-all --terragrunt-non-interactive --terragrunt-working-dir %s", environmentPath), &stdout, &stderr)
+	if err != nil {
+		return "", err
+	}
+
+	return stdout.String(), nil
+}
+
+func testTerragruntParallelismWithServer(t *testing.T, parallelism, timeBetweenBuckets, timeInsideBucket int) {
+	// setup a test server that sleeps for some time and sends the time
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(time.Duration(timeBetweenBuckets) * time.Second)
+		fmt.Fprintln(w, time.Now().Unix())
+	}))
+	defer ts.Close()
+	output, err := testRemoteFixtureParallelism(t, parallelism, ts.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	numberOfModules := 8
+
+	// parse output and sort the times
+	r := regexp.MustCompile(`out = (\d+)`)
+	matches := r.FindAllStringSubmatch(output, -1)
+	assert.True(t, len(matches) == numberOfModules) // 8 modules in the fixture
+	var times []int
+	for _, v := range matches {
+		match, _ := strconv.Atoi(v[1])
+		times = append(times, match)
+	}
+	sort.Slice(times, func(i, j int) bool {
+		return times[i] < times[j]
+	})
+	t.Logf("Parallelism test numberOfModules=%d p=%d timeBetweenBuckets=%d timeInsideBucket=%d sortedTimes=%v", numberOfModules, parallelism, timeBetweenBuckets, timeInsideBucket, times)
+
+	// there should be a difference of at least `timeBetweenBuckets` between items in the buckets
+	// there are 4 modules executed with parallelism 2, the time comparison is between [1, 3] and [2, 4]
+	// some scenarios
+	//
+	// time ->
+	//         1               3
+	//             2                4
+	//
+	// time ->
+	//         1                        3
+	//             2                4
+	for i := 0; i+parallelism < len(times); i += 1 {
+		j := i + parallelism
+		assert.True(t, times[j]-times[i] > timeBetweenBuckets)
+	}
+	// test within buckets, because times are sorted it's enough to compare the first and the last item in the bucket
+	// in the example above the items to compare would be [1, 2] [3, 4]
+	for i := 0; i < len(times); i += parallelism {
+		j := i + parallelism - 1
+		if j >= len(times) {
+			j = len(times) - 1
+		}
+		assert.True(t, times[j]-times[i] < timeInsideBucket)
+	}
+}
+
+func TestTerragruntParallelism(t *testing.T) {
```

Comment:
Unfortunately there are some operations done during apply-all that skew the times, for example I saw the following times for the case parallelism=4:

```
expected = 5, 5, 5, 5, 10, 10, 10, 10, 15, 15
got = [11 11 11 11 20 20 21 21 27 27]
got = [11 11 11 11 20 21 21 21 31 31]
got = [11 12 12 12 20 21 22 22 26 29]
```

To compute these times I calculated the time just before running `terragrunt apply-all` (start time) and for every time that output from a module I'd subtract the start time

```go
startTime = "2020-04-13T19:09:44"
```

I did an analysis of the logs to understand what additional operations are done:

The first log emitted is emitted 5 seconds after running terragrunt apply-all:

```
[terragrunt] [/var/folders/dn/2mqv6b716jq_xgxypnzpyr740000gq/T/terragrunt-test064579595/app5] 2020/04/13 19:09:47 Running command: terraform apply -input=false -auto-approve -var sleep_seconds=5
```

And the first time written to the output is at 11 seconds after the start time which explains the output

```
out = 2020-04-14T02:09:55Z
```

The total time is influenced by many factors whose duration we can't tell:

- The time to start terragrunt (reading all the modules recursively and preparing the data structures)
- The time to start a single module (2 seconds in the case below)
- The time for terraform to process the module and doing some operations on the state

```
[terragrunt] [/var/folders/dn/2mqv6b716jq_xgxypnzpyr740000gq/T/terragrunt-test064579595/app0] 2020/04/13 19:09:56 Running module /var/folders/dn/2mqv6b716jq_xgxypnzpyr740000gq/T/terragrunt-test064579595/app0 now
[terragrunt] [/var/folders/dn/2mqv6b716jq_xgxypnzpyr740000gq/T/terragrunt-test064579595/app0] 2020/04/13 19:09:56 Reading Terragrunt config file at /var/folders/dn/2mqv6b716jq_xgxypnzpyr740000gq/T/terragrunt-test064579595/app0/terragrunt.hcl
[terragrunt] [/var/folders/dn/2mqv6b716jq_xgxypnzpyr740000gq/T/terragrunt-test064579595/app0] 2020/04/13 19:09:56 Running command: terraform --version
[terragrunt] [/var/folders/dn/2mqv6b716jq_xgxypnzpyr740000gq/T/terragrunt-test064579595/app0] 2020/04/13 19:09:57 Running command: terraform apply -input=false -auto-approve -var sleep_seconds=5
```

As you see it's kinda hard to match the expected numbers with the numbers that we get however if we group them and make some calculations on top of the numbers there's some kind of pattern, in the test code that I wrote I tested two things, timeInsideBucket = 5, timeBetweenBuckets = 10, applied to the first output:

```
out =                [11 11 11 11] [20 20 21 21] [27 27]
timeInsideBucket                0             1       0 (all less than 5)       
timBetweenBuckets                 9             6       (all less than 10)
```

Reply:
I'm going to push my incomplete code so that you can check it too

Reply:
Gotcha. So it looks like we're dealing with 5-10 seconds of overhead on the individual `apply` calls... Could we work around it by setting `sleep_seconds` to `30` instead of `5`? And checking that the values are +/- 20?

Also, what is causing the 5-10 seconds of overhead? That seems abnormally high to me... Could some of it be the HTTP call `terraform` makes back to HashiCorp to check version numbers? You can disable that by setting the `CHECKPOINT_DISABLE` environment variable to `true`.

Reply:
@brikis98 thanks for the review, I came up with another strategy to test the timings match, the idea is to scale the timings reported by terraform by a constant factor, this factor is computed based on the timings' "width" or max reach which is simply the last item in the timings arrays because they're both sorted

For example for a case with parallelism set to 1 and sleeping for 5 seconds on every execution the expected timings are `expected=[5 10 15 20 25 30 35 40 45 50]` and the got times are `got=[7 15 23 31 38 45 53 60 68 75]` we see that the `width(expected) = 50` and `width(got) = 75` if we assume that these widths match then we have to map either one to the other through a scale operation.

If I apply the scale operation from got -> expected the scale factor is `k = width(got) / width(expected) = 75 / 50 = 1.5`, then each element in `got` is scaled by this factor yielding in the array `scaledGot=[4 10 15 20 25 30 35 40 45 50]` which is something that we can compare with the expected timings with more confidence

I've gotten these outputs with different parallelism parameters:

```
parallelism=1 sleep=5
expected=[5 10 15 20 25 30 35 40 45 50]
got=[7 15 23 31 38 45 53 60 68 75] 
scaledGot=[4 10 15 20 25 30 35 40 45 50]

parallelism=2 sleep=5
expected=[5 5 10 10 15 15 20 20 25 25]
got=[8 8 15 16 25 25 32 32 40 40]
scaledGot=[5 5 9 10 15 15 20 20 25 25]

parallelism=3 sleep=5
expected=[5 5 5 10 10 10 15 15 15 20]
got=[7 8 8 15 15 16 22 22 23 29]
scaledGot=[4 5 5 10 10 11 15 15 15 20]

parallelism=4 sleep=5
expected=[5 5 5 5 10 10 10 10 15 15]
got=[7 8 8 8 15 15 16 16 22 22]
scaledGot=[4 5 5 5 10 10 10 10 15 15]
```

Reply:
That seems like a reasonable approach, thanks!